src/store/RelayContext.js

```diff
@@ -0,0 +1,269 @@
+/**
+ * Copyright 2013-2015, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule RelayContext
+ * @typechecks
+ * @flow
+ */
+
+'use strict';
+
+const GraphQLFragmentPointer = require('GraphQLFragmentPointer');
+import type RelayMutation from 'RelayMutation';
+import type RelayMutationTransaction from 'RelayMutationTransaction';
+import type RelayQuery from 'RelayQuery';
+const RelayQueryResultObservable = require('RelayQueryResultObservable');
+const RelayStoreData = require('RelayStoreData');
+const RelayTaskScheduler = require('RelayTaskScheduler');
+
+const forEachRootCallArg = require('forEachRootCallArg');
+const invariant = require('invariant');
+const readRelayQueryData = require('readRelayQueryData');
+const warning = require('warning');
+
+import type {
+  Abortable,
+  Observable,
+  RelayMutationTransactionCommitCallbacks,
+  ReadyStateChangeCallback,
+  StoreReaderData,
+  StoreReaderOptions,
+} from 'RelayTypes';
+
+import type {
+  DataID,
+  RelayQuerySet,
+} from 'RelayInternalTypes';
+
+let defaultInstance;
+
+/**
+ * @public
+ *
+ * RelayContext is a caching layer that records GraphQL response data and enables
+ * resolving and subscribing to queries.
+ *
+ * === onReadyStateChange ===
+ *
+ * Whenever Relay sends a request for data via GraphQL, an "onReadyStateChange"
+ * callback can be supplied. This callback is called one or more times with a
+ * `readyState` object with the following properties:
+ *
+ *   aborted: Whether the request was aborted.
+ *   done: Whether all response data has been fetched.
+ *   error: An error in the event of a failure, or null if none.
+ *   ready: Whether the queries are at least partially resolvable.
+ *   stale: When resolvable during `forceFetch`, whether data is stale.
+ *
+ * If the callback is invoked with `aborted`, `done`, or a non-null `error`, the
+ * callback will never be called again. Example usage:
+ *
+ *  function onReadyStateChange(readyState) {
+ *    if (readyState.aborted) {
+ *      // Request was aborted.
+ *    } else if (readyState.error) {
+ *      // Failure occurred.
+ *    } else if (readyState.ready) {
+ *      // Queries are at least partially resolvable.
+ *      if (readyState.done) {
+ *        // Queries are completely resolvable.
+ *      }
+ *    }
+ *  }
+ *
+ */
+class RelayContext {
+  _storeData: RelayStoreData;
+
+  static getDefaultInstance(): RelayContext {
+    if (!defaultInstance) {
+      defaultInstance = new RelayContext(RelayStoreData.getDefaultInstance());
+    }
+    return defaultInstance;
+  }
+
+  constructor(storeData: RelayStoreData = new RelayStoreData()) {
```

Comment:
This type of optional with a default API seem easy to use but are too unpredictable - it's easy to forget to pass an instance somewhere and wonder why tests are failing. Let's require the instance to passed in for now (which you're already doing in tests).